File: 7.重建二叉树/7.重建二叉树.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # 返回构造的TreeNode根节点
    def reConstructBinaryTree(self, pre, tin):
        # write code here
        if len(pre)==0:
            return None
        root = TreeNode(pre[0])
        index = tin.index(root.val)
        leftNum = index
        rightNum = len(tin) - leftNum - 1
        root.left = self.reConstructBinaryTree(pre[1:1+leftNum],tin[0:leftNum])
        root.right = self.reConstructBinaryTree(pre[1+leftNum:],tin[leftNum+1:])
        return root

File: 7.重建二叉树/7.重建二叉树.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
       
        if (pre.size()==0)
            return NULL;
        TreeNode* root = new TreeNode(pre[0]);
        int rootIndex = 0;
        while(vin[rootIndex]!=root->val)
            rootIndex++;
        int leftNum = rootIndex;
        int rightNum = pre.size()-leftNum-1;
        root->left = reConstructBinaryTree(vector<int>(pre.begin()+1,pre.begin()+1+leftNum),vector<int>(vin.begin(),vin.begin()+leftNum));
        root->right = reConstructBinaryTree(vector<int>(pre.begin()+1+leftNum,pre.end()),vector<int>(vin.begin()+leftNum+1,vin.end())); 
        return root;
        
    }
};